### DISCUSSION TITLE
tryWithTrashed or secureWithTrashed

### DISCUSSION THREAD

**Discussion by kobe-ra** [laravel/framework#51658] [Ideas]
hi, i've been using laravel professionally for ~ 5 years now, so i would love to contribute, but so far i don't know how. if anyone would share some tips i'd be grateful.
code like `$data['fileable_type']::withTrashed()->get()` will error, if the class doesn't use `SoftDeletes`.

my proposition is to add a method `tryWithTrashed` or similar name that would look like an improved version of this
```
if (in_array(SoftDeletes::class, class_uses_recursive($this))) {
    return $q->withTrashed();
}
return $q;
```

now i found that `withTrashed` method comes from `addWithTrashed` in `Database/Eloquent/SoftDeletingScope.php`, but the new method couldn't be added there, since model wouldn't have access to it.

where could it be placed? i was lookin at places like `Database/Query/Builder.php` or `Database/Eloquent/Builder.php` but i'm not sure how to test code i'd add there.

any help is appreciated.
thanks.

**Comment by henzeb** [comment]
I doubt they will merge this... but something like this should go in Model or in a trait. In your case I suggest creating a helper method.

```php
tryWithTrashed($data['fileable_type'])->get();
```

The reason why I think they won't merge this is that they would have to make an exception for all possible situations. `withoutTrash` ,  `onlyTrashed` etc, and that's just the trash related ones. You have to come up with a more generic solution.

**Reply by kobe-ra** [reply]
that's a good point. with this in mind, i would imagine a method like `safeSoftDeletes(fn)`, so we could call `safeSoftDeletes(withTrashed())`, or `safeSoftDeletes(onlyTrashed())`, and the method would call its argument if `SoftDeletes` is used on that model.
as for my own project - i have this solved with a helper, i was just lookin for something cleaner.
do you think this could work?
